File: ykpsmuttauth/ykpsmuttauth.go
/*
 * Copyright (c) 2024 Runxi Yu <https://runxiyu.org>
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Key inspiration was taken from mutt_oauth2.py written by Alexander Perlis,
 * licensed under the GNU General Public License, version 2 or later, as
 * published by the Free Software Foundation. I don't think that this program
 * is a derivative work of mutt_oauth2.py in terms of GPL interpretation, but I
 * might be wrong.  Consult your lawyer if you want to use this program in a
 * context incompatible with the GPL.
 * https://raw.githubusercontent.com/muttmua/mutt/master/contrib/mutt_oauth2.py
 */

package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"time"
)

type Registration struct {
	AuthorizeEndpoint  string
	DevicecodeEndpoint string
	TokenEndpoint      string
	Tenant             string
	Scope              string
	ClientID           string
}

type Token struct {
	AccessToken           string `json:"access_token"`
	AccessTokenExpiration string `json:"access_token_expiration"`
	RefreshToken          string `json:"refresh_token"`
	Email                 string `json:"email"`
}

var ykps = Registration{
	AuthorizeEndpoint:  "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/authorize",
	DevicecodeEndpoint: "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/devicecode",
	TokenEndpoint:      "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/token",
	Tenant:             "ddd3d26c-b197-4d00-a32d-1ffd84c0c295",
	Scope:              "offline_access https://outlook.office.com/IMAP.AccessAsUser.All https://outlook.office.com/SMTP.Send",
	ClientID:           "fea760d5-b496-4f63-be1e-93855c1c5f78",
}

var token Token
var tokenFile string
var authorize_email string

func init() {
	flag.StringVar(&tokenFile, "tokenfile", "", "(required) persistent token storage")
	flag.StringVar(&authorize_email, "authorize", "", "email to newly authorize")
	flag.Parse()
}

func readTokenFile() error {
	data, err := os.ReadFile(tokenFile)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, &token)
}

func writeTokenFile() error {
	data, err := json.Marshal(&token)
	if err != nil {
		return err
	}
	return os.WriteFile(tokenFile, data, 0600)
}

func accessTokenValid() bool {
	if token.AccessTokenExpiration == "" {
		return false
	}
	expirationTime, err := time.Parse(time.RFC3339, token.AccessTokenExpiration)
	if err != nil {
		return false
	}
	return time.Now().Before(expirationTime)
}

func updateTokens(r map[string]interface{}) error {
	token.AccessToken = r["access_token"].(string)
	expiresIn := int(r["expires_in"].(float64))
	token.AccessTokenExpiration = time.Now().Add(time.Duration(expiresIn) * time.Second).Format(time.RFC3339)
	if refreshToken, ok := r["refresh_token"]; ok {
		token.RefreshToken = refreshToken.(string)
	}
	return writeTokenFile()
}

func generateCodeVerifier() (string, string, error) {
	verifier := make([]byte, 90)
	_, err := rand.Read(verifier)
	if err != nil {
		return "", "", err
	}
	codeVerifier := base64.RawURLEncoding.EncodeToString(verifier)
	hash := sha256.Sum256([]byte(codeVerifier))
	codeChallenge := base64.RawURLEncoding.EncodeToString(hash[:])
	return codeVerifier, codeChallenge, nil
}

func startHTTPServer() (chan string, int, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return nil, 0, err
	}

	codeChan := make(chan string)
	server := &http.Server{}
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		code := r.URL.Query().Get("code")
		if code != "" {
			codeChan <- code
		}
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusOK)
		fmt.Fprintf(w, "Done, please check your terminal")
	})

	go func() {
		if err := server.Serve(listener); err != http.ErrServerClosed {
			log.Fatalf("HTTP server error: %v", err)
		}
	}()

	return codeChan, listener.Addr().(*net.TCPAddr).Port, nil
}

func authorize_request() error {
	p := url.Values{
		"client_id": {ykps.ClientID},
		"tenant":    {ykps.Tenant},
		"scope":     {ykps.Scope},
	}
	codeVerifier, codeChallenge, err := generateCodeVerifier()
	if err != nil {
		return err
	}
	codeChan, port, err := startHTTPServer()
	if err != nil {
		return err
	}
	redirectURI := fmt.Sprintf("http://localhost:%d/", port)
	p.Set("redirect_uri", redirectURI)
	p.Set("response_type", "code")
	p.Set("code_challenge", codeChallenge)
	p.Set("code_challenge_method", "S256")
	p.Set("login_hint", token.Email)
	authorizeURL := ykps.AuthorizeEndpoint + "?" + p.Encode()
	fmt.Fprintf(os.Stderr, "Please visit:\n%s\n", authorizeURL)

	authCode := <-codeChan

	data := url.Values{
		"client_id":     {ykps.ClientID},
		"tenant":        {ykps.Tenant},
		"code":          {authCode},
		"redirect_uri":  {redirectURI},
		"grant_type":    {"authorization_code"},
		"code_verifier": {codeVerifier},
	}

	resp, err := http.PostForm(ykps.TokenEndpoint, data)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return errors.New(resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var response map[string]interface{}
	if err := json.Unmarshal(body, &response); err != nil {
		return err
	}

	if _, ok := response["error"]; ok {
		return fmt.Errorf("authorization error: %v", response["error"])
	}

	return updateTokens(response)
}

func refreshToken() error {
	if token.RefreshToken == "" {
		return errors.New("no refresh token available")
	}

	data := url.Values{
		"client_id":     {ykps.ClientID},
		"tenant":        {ykps.Tenant},
		"refresh_token": {token.RefreshToken},
		"grant_type":    {"refresh_token"},
	}

	resp, err := http.PostForm(ykps.TokenEndpoint, data)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return errors.New(resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var response map[string]interface{}
	if err := json.Unmarshal(body, &response); err != nil {
		return err
	}

	if _, ok := response["error"]; ok {
		return fmt.Errorf("refresh token error: %v", response["error"])
	}

	return updateTokens(response)
}

func main() {
	if tokenFile == "" {
		log.Fatal("You must provide a token file path.")
	}

	if err := readTokenFile(); err != nil {
		if !os.IsNotExist(err) {
			log.Fatalf("Error reading token file: %v", err)
		}
		if authorize_email == "" {
			log.Fatalf("You must run the script with --authorize at least once.")
		}
		token.Email = authorize_email
		token.AccessToken = ""
		token.AccessTokenExpiration = ""
		token.RefreshToken = ""
		if err := writeTokenFile(); err != nil {
			log.Fatalf("Error writing token file: %v", err)
		}
	}

	if authorize_email != "" {
		if err := authorize_request(); err != nil {
			log.Fatalf("Authorization error: %v", err)
		}
	} else if !accessTokenValid() {
		if err := refreshToken(); err != nil {
			log.Fatalf("Refresh token error: %v", err)
		}
	}

	if !accessTokenValid() {
		log.Fatal("No valid access token. This should not be able to happen.")
	}

	fmt.Println(token.AccessToken)
}

File: ykpsmuttauth/ykpsmuttauth.py
#!/usr/bin/env python3
#
# Mutt OAuth2 token management script, version 2020-08-07
# Written against python 3.7.3, not tried with earlier python versions.
# Modified for YK Pao School application registrations.
#
#   Copyright (C) 2020 Alexander Perlis
#   Copyright (C) 2024 Runxi Yu <me@runxiyu.org>
#
#   This program is free software; you can redistribute it and/or
#   modify it under the terms of the GNU General Public License as
#   published by the Free Software Foundation; either version 2 of the
#   License, or (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
#   02110-1301, USA.

import sys
import json
import argparse
import urllib.parse
import urllib.request
import imaplib
import smtplib
import base64
import secrets
import hashlib
import time
from datetime import timedelta, datetime
from pathlib import Path
import socket
import http.server
import subprocess
import readline


registrations = {
    "ykps": {
        "authorize_endpoint": "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/authorize",
        "devicecode_endpoint": "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/devicecode",
        "token_endpoint": "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/v2.0/token",
        "redirect_uri": "https://login.microsoftonline.com/ddd3d26c-b197-4d00-a32d-1ffd84c0c295/oauth2/nativeclient",
        "tenant": "ddd3d26c-b197-4d00-a32d-1ffd84c0c295",
        "imap_endpoint": "outlook.office365.com",
        "smtp_endpoint": "smtp.office365.com",
        "sasl_method": "XOAUTH2",
        "scope": (
            "offline_access https://outlook.office.com/IMAP.AccessAsUser.All "
            "https://outlook.office.com/SMTP.Send"
        ),
        "client_id": "fea760d5-b496-4f63-be1e-93855c1c5f78",
        "client_secret": "",
    },
}

ap = argparse.ArgumentParser()
ap.add_argument("tokenfile", help="persistent token storage")
ap.add_argument(
    "-a", "--authorize", action="store_true", help="manually authorize new tokens"
)
args = ap.parse_args()

token = {}
path = Path(args.tokenfile)
if path.exists():
    token = json.loads(path.read_bytes())


def writetokenfile():
    """Writes global token dictionary into token file."""
    if not path.exists():
        path.touch(mode=0o600)
    path.write_bytes(json.dumps(token).encode())


print("Obtained from token file:", json.dumps(token), file=sys.stderr)
if not token:
    if not args.authorize:
        sys.exit('You must run script with "--authorize" at least once.')
    print("Available app and endpoint registrations:", *registrations, file=sys.stderr)
    token["registration"] = input("OAuth2 registration: ")
    token["email"] = input("Account e-mail address: ")
    token["access_token"] = ""
    token["access_token_expiration"] = ""
    token["refresh_token"] = ""
    writetokenfile()

if token["registration"] not in registrations:
    sys.exit(
        f'ERROR: Unknown registration "{token["registration"]}". Delete token file '
        f"and start over."
    )
registration = registrations[token["registration"]]

authflow = "localhostauthcode"

baseparams = {"client_id": registration["client_id"], "tenant": registration["tenant"]}


def access_token_valid():
    """Returns True when stored access token exists and is still valid at this time."""
    token_exp = token["access_token_expiration"]
    return token_exp and datetime.now() < datetime.fromisoformat(token_exp)


def update_tokens(r):
    """Takes a response dictionary, extracts tokens out of it, and updates token file."""
    token["access_token"] = r["access_token"]
    token["access_token_expiration"] = (
        datetime.now() + timedelta(seconds=int(r["expires_in"]))
    ).isoformat()
    if "refresh_token" in r:
        token["refresh_token"] = r["refresh_token"]
    writetokenfile()
    if args.verbose:
        print(
            f'NOTICE: Obtained new access token, expires {token["access_token_expiration"]}.',
            file=sys.stderr,
        )


if args.authorize:
    p = baseparams.copy()
    p["scope"] = registration["scope"]

    verifier = secrets.token_urlsafe(90)
    challenge = base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest())[
        :-1
    ]
    redirect_uri = registration["redirect_uri"]
    listen_port = 0

    # Find an available port to listen on
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    listen_port = s.getsockname()[1]
    s.close()
    redirect_uri = "http://localhost:" + str(listen_port) + "/"
    # Probably should edit the port number into the actual redirect URL.

    p.update(
        {
            "login_hint": token["email"],
            "response_type": "code",
            "redirect_uri": redirect_uri,
            "code_challenge": challenge,
            "code_challenge_method": "S256",
        }
    )
    print(
        registration["authorize_endpoint"]
        + "?"
        + urllib.parse.urlencode(p, quote_via=urllib.parse.quote),
        file=sys.stderr,
    )

    authcode = ""
    print(
        "Visit displayed URL to authorize this application. Waiting...",
        end="",
        flush=True,
        file=sys.stderr,
    )

    class MyHandler(http.server.BaseHTTPRequestHandler):

        def do_HEAD(self):
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()

        def do_GET(self):
            global authcode
            querystring = urllib.parse.urlparse(self.path).query
            querydict = urllib.parse.parse_qs(querystring)
            if "code" in querydict:
                authcode = querydict["code"][0]
            self.do_HEAD()
            self.wfile.write(b"<html><head><title>Authorizaton result</title></head>")
            self.wfile.write(
                b"<body><p>Authorization redirect completed. You may "
                b"close this window.</p></body></html>"
            )

    with http.server.HTTPServer(("127.0.0.1", listen_port), MyHandler) as httpd:
        try:
            httpd.handle_request()
        except KeyboardInterrupt:
            pass

    if not authcode:
        sys.exit("Did not obtain an authcode.")

    for k in (
        "response_type",
        "login_hint",
        "code_challenge",
        "code_challenge_method",
    ):
        del p[k]
    p.update(
        {
            "grant_type": "authorization_code",
            "code": authcode,
            "client_secret": registration["client_secret"],
            "code_verifier": verifier,
        }
    )
    print("Exchanging the authorization code for an access token", file=sys.stderr)
    try:
        response = urllib.request.urlopen(
            registration["token_endpoint"], urllib.parse.urlencode(p).encode()
        )
    except urllib.error.HTTPError as err:
        print(err.code, err.reason, file=sys.stderr)
        response = err
    response = response.read()
    print(response, file=sys.stderr)
    response = json.loads(response)
    if "error" in response:
        print(response["error"], file=sys.stderr)
        if "error_description" in response:
            print(response["error_description"], file=sys.stderr)
        sys.exit(1)

    update_tokens(response)


if not access_token_valid():
    if args.verbose:
        print(
            "NOTICE: Invalid or expired access token; using refresh token "
            "to obtain new access token.",
            file=sys.stderr,
        )
    if not token["refresh_token"]:
        sys.exit('ERROR: No refresh token. Run script with "--authorize".')
    p = baseparams.copy()
    p.update(
        {
            "client_secret": registration["client_secret"],
            "refresh_token": token["refresh_token"],
            "grant_type": "refresh_token",
        }
    )
    try:
        response = urllib.request.urlopen(
            registration["token_endpoint"], urllib.parse.urlencode(p).encode()
        )
    except urllib.error.HTTPError as err:
        print(err.code, err.reason, file=sys.stderr)
        response = err
    response = response.read()
    print(response, file=sys.stderr)
    response = json.loads(response)
    if "error" in response:
        print(response["error"], file=sys.stderr)
        if "error_description" in response:
            print(response["error_description"], file=sys.stderr)
        print('Perhaps refresh token invalid. Try running once with "--authorize"')
        sys.exit(1)
    update_tokens(response)


if not access_token_valid():
    sys.exit("ERROR: No valid access token. This should not be able to happen.")


# print("Access Token: ", end="", file=sys.stderr)
print(token["access_token"])


def build_sasl_string(user, host, port, bearer_token):
    return f"user={user}\1auth=Bearer {bearer_token}\1\1"

